### PR TITLE
 WIP: Fix "Sub-directory Support"

### PR DESCRIPTION
**Pull request by mpikzink**
This changes works im my Apache env.

1. The Double prefix from #12205 @candlerb is fixed with the changes in Url.php

2. I had several problems with leading "/" in href.

3. The root-Level URL from #12205 is fixed with web.php

Please give a short description what your pull request is for

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by ThomasCr** [comment]
Hi, there is still a Bug with "Sub-directory Support" after your changes.
When I go to Ports view and select "All-Ports" and hower on tehat page over the port col the grafs are not shown.

grafik

File: resources/views/locations.blade.php [hunk 1]
```diff
@@ -87,7 +87,7 @@
                 url: "{{ url('/ajax/table/location') }}",
```
Comment (ThomasCr) [inline resources/views/locations.blade.php:87]: /ajax looks also hard-coded

**Comment by mpikzink** [comment]
> Hi, there is still a Bug with "Sub-directory Support" after your changes.
> When I go to **Ports** view and select **All-Ports** and hower on that page over the **Port** column the grafs are not shown.
> 
> grafik
> 
> Same thing on **Overview** -> **IPv4 Address**, when I hover over the Interfaces
> 
> grafik
> 
> There is the subdirectory lost. - All other pages / graphs (also the mouse over ones) are working.
> 
> Also the double /librenms/librenms is still present on devices and ports view...? Do we need to flush caches somewhere?
> What I already tested is: `rm storage/framework/views/*.php`
> 
> I found also this with: `grep -R -e '/devices' -e '/ports' . | grep -v -e routes-v7 -e mib -e '/doc'` from the install directory
> 
> ```
> diff --git a/includes/html/dev-groups-overview-data.inc.php b/includes/html/dev-groups-overview-data.inc.php
> index 4a450ac0f..1cc23f603 100644
> --- a/includes/html/dev-groups-overview-data.inc.php
> +++ b/includes/html/dev-groups-overview-data.inc.php
> @@ -18,7 +18,7 @@ if (count($device_groups)) {
>                          <div class="col-sm-12">
>                          <?php foreach ($device_groups as $group) { ?>
>                              <span style="margin: 8px;">
> -                                <a href="/devices/group=<?=$group['id']?>" target="_blank"><?=$group['name']?></a>
> +                                <a href="devices/group=<?=$group['id']?>" target="_blank"><?=$group['name']?></a>
>                              </span>
>                          <?php } ?>
>                          </div>
> ```
> 
> there are also some more things in api.php and BasicApiTest.php which are also not looking so well.

**Comment by murrant** [comment]
Can you add `dd(url()->current(), url('/'), url(''));` to a page and tell me what it looks like?

**Comment by mpikzink** [comment]
Screenshot_2020-12-16_16-20-45
I think url() / Laravel is not the problem. I think the problem coms just with hard coded links in html. My last commits didn't change thinks in my setup.

**Comment by mpikzink** [comment]
Subdir_ok
Without leading / everything is fine.

**Comment by murrant** [comment]
So, seems like the fix would be to convert all urls to use the url() helper (route() is preferred for full laravel pages).

I was curious because I'm pretty sure the leading slash is preferred when using the url() helper.

**Comment by mpikzink** [comment]
Subdir_missing
But with leading / the subdirectory will not be included!?

**Comment by murrant** [comment]
Only when using raw urls.  As shown by your output the url helper knows what to prepend to the url correctly.  It should not matter either way because the url helper will put a slash in between the root and path every time.

**Comment by mpikzink** [comment]
> So, seems like the fix would be to convert all urls to use the url() helper (route() is preferred for full laravel pages).
> 
> I was curious because I'm pretty sure the leading slash is preferred when using the url() helper.

OK, I can change the hrefs i found to route(/...) but how to use route() in page which are not converted to Laravel blade? Can I think I can't use {{ }}directly.

**Comment by murrant** [comment]
Look at the name column from this command `php artisan route:list` to see all routes that can use the route helper.  Then call like this for example `route('device-groups.index')`

For others, just wrap the url() function around the url.

`{{ url('blah') }}` is only in blade templates and is almost the same as `<?php echo url('blah'); ?>` in php files.

**Comment by mpikzink** [comment]
OK, thanks.  Can you show me where the popover links are created? I searched for it but either I can not find the place or something is cached and I do not know how to renew it.

**Comment by murrant** [comment]
generate_port_link() has not been ported to use the url helper yet...

File: LibreNMS/Util/Url.php [hunk 2]
```diff
@@ -225,13 +225,7 @@ public static function sensorLink($sensor, $text = null, $type = null, $overlib
...
     public static function deviceUrl($device, $vars = [])
     {
-        $routeParams = [is_int($device) ? $device : $device->device_id];
-        if (isset($vars['tab'])) {
-            $routeParams[] = $vars['tab'];
-            unset($vars['tab']);
-        }
-
-        return route('device', $routeParams) . self::urlParams($vars);
+        return self::generate(['page' => 'device', 'device' => is_int($device) ? $device : $device->device_id], $vars);
```
Comment (murrant) [inline LibreNMS/Util/Url.php:228]: This is going to cause issues, above code has a workaround for a bug with legacy code calling it.

**Comment by murrant** [comment]
@mpikzink I suggest smaller changes across multiple pull requests.  Will be much easier to make progress.

**Comment by murrant** [comment]
Closing due to inactivity, feel free to reopen.

**Comment by arrmo** [comment]
FYI, I definitely do still have this issue (just to let you know). The "hack" I have in place here works fine ... but I'm not sure it's ready for prime time. I don't see it causing other issues, but I may not be seeing some of the side effects. That's why I haven't tried pushing it (PR) ... but I can, no issue here. Thoughts?

Thanks!

**Comment by murrant** [comment]
@arrmo issues are known. One problem is this PR is too big so a few issues held up known things that do fix stuff.  If you have the time, make multiple smaller PRs to fix issues.

**Pull request closed without merging**